File: python/phasesmith/quantitative.py
"""Quantitative phase analysis from compatible refined phase scales."""

from __future__ import annotations

import math
from collections.abc import Iterable
from dataclasses import dataclass

import numpy as np
from numpy.typing import ArrayLike, NDArray


def _positive_vector(
    values: ArrayLike, name: str, *, allow_zero: bool = False
) -> NDArray[np.float64]:
    array = np.asarray(values, dtype=np.float64)
    if array.ndim != 1 or array.size == 0:
        raise ValueError(f"{name} must be a non-empty one-dimensional array")
    if not np.isfinite(array).all():
        raise ValueError(f"{name} must contain only finite values")
    if allow_zero:
        if np.any(array < 0.0):
            raise ValueError(f"{name} must be nonnegative")
    elif np.any(array <= 0.0):
        raise ValueError(f"{name} must be positive")
    return array


def weight_fractions_from_scale(
    scale: ArrayLike,
    formula_units_per_cell: ArrayLike,
    formula_mass_g_mol: ArrayLike,
    cell_volume_angstrom3: ArrayLike,
) -> NDArray[np.float64]:
    r"""Convert compatible Rietveld scales to crystalline weight fractions.

    This implements the Hill--Howard relation

    ``W_p = S_p (Z M V)_p / sum_i S_i (Z M V)_i``.

    All scales must use the same structure-factor and intensity normalization.
    The result describes only the supplied crystalline phases; it does not
    measure unidentified or amorphous material without an appropriate standard.
    """

    scales = _positive_vector(scale, "scale", allow_zero=True)
    z = _positive_vector(formula_units_per_cell, "formula_units_per_cell")
    mass = _positive_vector(formula_mass_g_mol, "formula_mass_g_mol")
    volume = _positive_vector(cell_volume_angstrom3, "cell_volume_angstrom3")
    if not (scales.shape == z.shape == mass.shape == volume.shape):
        raise ValueError("all quantitative-phase arrays must have the same shape")
    with np.errstate(over="ignore", invalid="ignore"):
        contributions = scales * z * mass * volume
    if not np.isfinite(contributions).all():
        raise ValueError("quantitative-phase contributions overflowed")
    total = float(np.sum(contributions))
    if not math.isfinite(total) or total <= 0.0:
        raise ValueError("at least one phase scale must be positive")
    fractions = np.ascontiguousarray(contributions / total)
    fractions.flags.writeable = False
    return fractions


@dataclass(frozen=True, slots=True)
class QuantitativePhase:
    """Metadata required to interpret one compatible refined phase scale."""

    phase_id: str
    scale: float
    formula_units_per_cell: float
    formula_mass_g_mol: float
    cell_volume_angstrom3: float

    def __post_init__(self) -> None:
        if not isinstance(self.phase_id, str) or not self.phase_id:
            raise ValueError("phase_id must be a non-empty string")
        values = (
            self.scale,
            self.formula_units_per_cell,
            self.formula_mass_g_mol,
            self.cell_volume_angstrom3,
        )
        if any(isinstance(value, bool) or not isinstance(value, (int, float)) for value in values):
            raise TypeError("quantitative-phase numeric fields must be real scalars")
        if not all(math.isfinite(float(value)) for value in values):
            raise ValueError("quantitative-phase numeric fields must be finite")
        if self.scale < 0.0:
            raise ValueError("phase scale must be nonnegative")
        if (
            min(
                self.formula_units_per_cell,
                self.formula_mass_g_mol,
                self.cell_volume_angstrom3,
            )
            <= 0.0
        ):
            raise ValueError("Z, formula mass, and cell volume must be positive")


@dataclass(frozen=True, slots=True)
class PhaseWeightFraction:
    """One labeled normalized result from crystalline quantitative analysis."""

    phase_id: str
    weight_fraction: float

    def __post_init__(self) -> None:
        if not isinstance(self.phase_id, str) or not self.phase_id:
            raise ValueError("phase_id must be a non-empty string")
        if not math.isfinite(self.weight_fraction) or not 0.0 <= self.weight_fraction <= 1.0:
            raise ValueError("weight_fraction must be finite and between zero and one")


def quantitative_phase_analysis(
    phases: Iterable[QuantitativePhase],
) -> tuple[PhaseWeightFraction, ...]:
    """Return labeled weight fractions while preserving caller phase order."""

    records = tuple(phases)
    if not records:
        raise ValueError("quantitative phase analysis requires at least one phase")
    if any(not isinstance(phase, QuantitativePhase) for phase in records):
        raise TypeError("phases must contain QuantitativePhase values")
    phase_ids = tuple(phase.phase_id for phase in records)
    if len(set(phase_ids)) != len(phase_ids):
        raise ValueError("quantitative phase IDs must be unique")
    fractions = weight_fractions_from_scale(
        [phase.scale for phase in records],
        [phase.formula_units_per_cell for phase in records],
        [phase.formula_mass_g_mol for phase in records],
        [phase.cell_volume_angstrom3 for phase in records],
    )
    return tuple(
        PhaseWeightFraction(phase_id, float(fraction))
        for phase_id, fraction in zip(phase_ids, fractions, strict=True)
    )

File: crates/phasesmith-workflows/src/quantitative.rs
//! Quantitative crystalline phase fractions from compatible Rietveld scales.

use std::error::Error;
use std::fmt::{Display, Formatter};

/// Metadata required by the Hill--Howard scale relation.
#[derive(Clone, Debug, PartialEq)]
pub struct QuantitativePhase {
    /// Stable phase identity.
    pub phase_id: String,
    /// Compatible non-negative refined phase scale.
    pub scale: f64,
    /// Formula units per crystallographic unit cell.
    pub formula_units_per_cell: f64,
    /// Formula mass in grams per mole.
    pub formula_mass_g_mol: f64,
    /// Unit-cell volume in cubic ångströms.
    pub cell_volume_angstrom3: f64,
}

impl QuantitativePhase {
    /// Validate one quantitative phase record.
    ///
    /// # Errors
    ///
    /// Returns [`QuantitativeError`] for empty identity, negative scale, or
    /// non-positive/non-finite physical metadata.
    pub fn new(
        phase_id: impl Into<String>,
        scale: f64,
        formula_units_per_cell: f64,
        formula_mass_g_mol: f64,
        cell_volume_angstrom3: f64,
    ) -> Result<Self, QuantitativeError> {
        let result = Self {
            phase_id: phase_id.into(),
            scale,
            formula_units_per_cell,
            formula_mass_g_mol,
            cell_volume_angstrom3,
        };
        result.validate()?;
        Ok(result)
    }

    fn validate(&self) -> Result<(), QuantitativeError> {
        if self.phase_id.is_empty() {
            return Err(QuantitativeError::EmptyPhaseId);
        }
        if !self.scale.is_finite() || self.scale < 0.0 {
            return Err(QuantitativeError::InvalidScale);
        }
        if [
            self.formula_units_per_cell,
            self.formula_mass_g_mol,
            self.cell_volume_angstrom3,
        ]
        .into_iter()
        .any(|value| !value.is_finite() || value <= 0.0)
        {
            return Err(QuantitativeError::InvalidMetadata);
        }
        Ok(())
    }
}

/// One normalized crystalline weight fraction.
#[derive(Clone, Debug, PartialEq)]
pub struct PhaseWeightFraction {
    /// Stable phase identity.
    pub phase_id: String,
    /// Fraction of the supplied crystalline phases in `[0, 1]`.
    pub weight_fraction: f64,
}

/// Weight fractions and their row-major covariance propagated from phase scales.
#[derive(Clone, Debug, PartialEq)]
pub struct QuantitativePhaseAnalysis {
    /// Normalized crystalline weight fractions in input order.
    pub phases: Vec<PhaseWeightFraction>,
    /// Row-major fraction covariance with dimension `phases.len()`.
    pub covariance: Vec<f64>,
}

/// Calculate Hill--Howard crystalline weight fractions in input order.
///
/// `W_p = S_p (Z M V)_p / sum_i S_i (Z M V)_i`.
///
/// # Errors
///
/// Returns [`QuantitativeError`] for empty, duplicate, invalid, overflowing,
/// or all-zero phase records.
pub fn quantitative_phase_analysis(
    phases: &[QuantitativePhase],
) -> Result<Vec<PhaseWeightFraction>, QuantitativeError> {
    if phases.is_empty() {
        return Err(QuantitativeError::EmptyPhases);
    }
    let mut identities = std::collections::BTreeSet::new();
    let mut contributions = Vec::with_capacity(phases.len());
    for phase in phases {
        phase.validate()?;
        if !identities.insert(&phase.phase_id) {
            return Err(QuantitativeError::DuplicatePhaseId);
        }
        let contribution = phase.scale
            * phase.formula_units_per_cell
            * phase.formula_mass_g_mol
            * phase.cell_volume_angstrom3;
        if !contribution.is_finite() {
            return Err(QuantitativeError::ContributionOverflow);
        }
        contributions.push(contribution);
    }
    let total = contributions.iter().sum::<f64>();
    if !total.is_finite() || total <= 0.0 {
        return Err(QuantitativeError::ZeroTotal);
    }
    Ok(phases
        .iter()
        .zip(contributions)
        .map(|(phase, contribution)| PhaseWeightFraction {
            phase_id: phase.phase_id.clone(),
            weight_fraction: contribution / total,
        })
        .collect())
}

/// Calculate weight fractions and analytically propagate a phase-scale covariance.
///
/// For `c_i = S_i (Z M V)_i`, `W_i = c_i / sum(c)`, the scale derivative is
/// `dW_i/dS_j = (delta_ij k_i - W_i k_j) / sum(c)`, where `k_i = (Z M V)_i`.
///
/// # Errors
///
/// Returns [`QuantitativeError`] for invalid phases or covariance shape/values.
pub fn quantitative_phase_analysis_with_covariance(
    phases: &[QuantitativePhase],
    scale_covariance: &[f64],
) -> Result<QuantitativePhaseAnalysis, QuantitativeError> {
    let fractions = quantitative_phase_analysis(phases)?;
    let count = phases.len();
    if scale_covariance.len()
        != count
            .checked_mul(count)
            .ok_or(QuantitativeError::CovarianceShape)?
    {
        return Err(QuantitativeError::CovarianceShape);
    }
    if scale_covariance.iter().any(|value| !value.is_finite()) {
        return Err(QuantitativeError::NonFiniteCovariance);
    }
    let factors = phases
        .iter()
        .map(|phase| {
            phase.formula_units_per_cell * phase.formula_mass_g_mol * phase.cell_volume_angstrom3
        })
        .collect::<Vec<_>>();
    let total = phases
        .iter()
        .zip(&factors)
        .map(|(phase, factor)| phase.scale * factor)
        .sum::<f64>();
    let mut jacobian = vec![0.0; count * count];
    for row in 0..count {
        for column in 0..count {
            jacobian[row * count + column] = (if row == column { factors[row] } else { 0.0 }
                - fractions[row].weight_fraction * factors[column])
                / total;
        }
    }
    let mut covariance = vec![0.0; count * count];
    for row in 0..count {
        for column in 0..count {
            let mut value = 0.0;
            for left in 0..count {
                for right in 0..count {
                    value += jacobian[row * count + left]
                        * scale_covariance[left * count + right]
                        * jacobian[column * count + right];
                }
            }
            if !value.is_finite() {
                return Err(QuantitativeError::NonFiniteCovariance);
            }
            covariance[row * count + column] = value;
        }
    }
    Ok(QuantitativePhaseAnalysis {
        phases: fractions,
        covariance,
    })
}

/// Invalid quantitative-phase input.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum QuantitativeError {
    /// At least one phase is required.
    EmptyPhases,
    /// Phase identity must not be empty.
    EmptyPhaseId,
    /// Phase identities must be unique.
    DuplicatePhaseId,
    /// Scale must be finite and non-negative.
    InvalidScale,
    /// Z, mass, and cell volume must be positive and finite.
    InvalidMetadata,
    /// A scale-times-metadata product overflowed.
    ContributionOverflow,
    /// At least one phase scale must be positive.
    ZeroTotal,
    /// Phase-scale covariance did not have square phase dimension.
    CovarianceShape,
    /// Phase-scale covariance or propagated covariance was not finite.
    NonFiniteCovariance,
}

impl Display for QuantitativeError {
    fn fmt(&self, formatter: &mut Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(match self {
            Self::EmptyPhases => "quantitative phase analysis requires at least one phase",
            Self::EmptyPhaseId => "quantitative phase IDs must not be empty",
            Self::DuplicatePhaseId => "quantitative phase IDs must be unique",
            Self::InvalidScale => "quantitative phase scales must be finite and non-negative",
            Self::InvalidMetadata => "quantitative Z, mass, and volume must be positive and finite",
            Self::ContributionOverflow => "quantitative phase contribution overflowed",
            Self::ZeroTotal => "at least one quantitative phase scale must be positive",
            Self::CovarianceShape => "scale covariance must be square with one row per phase",
            Self::NonFiniteCovariance => {
                "scale covariance and propagated covariance must be finite"
            }
        })
    }
}

impl Error for QuantitativeError {}
